### PR TITLE
[ide] Added cancel indicator to new semantic highlighting API

### PR DESCRIPTION
**Pull request by spoenemann**
This extension of the API would allow better performance in the web context, especially if computing semantic highlighting is expensive.

### CODE REVIEW

File: plugins/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/syntaxcoloring/DefaultSemanticHighlightingCalculator.java [hunk 1]
```diff
@@ -63,22 +69,26 @@ public void provideHighlightingFor(XtextResource resource, IHighlightedPositionA
...
-		highlightElementRecursively(element, acceptor);
+		highlightElementRecursively(element, acceptor, cancelIndicator);
 	}
 
-	protected void highlightElementRecursively(EObject element, IHighlightedPositionAcceptor acceptor) {
+	protected void highlightElementRecursively(EObject element, IHighlightedPositionAcceptor acceptor,
+			CancelIndicator cancelIndicator) {
 		TreeIterator<EObject> iterator = EcoreUtil2.eAll(element);
 		while (iterator.hasNext()) {
+			operationCanceledManager.checkCanceled(cancelIndicator);
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/syntaxcoloring/DefaultSemanticHighlightingCalculator.java]: This appears to be very fine-grained. Maybe we could simply check on each acceptor.accept call by default? Or every 1000 steps?

Comment (szarnekow) [inline plugins/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/syntaxcoloring/DefaultSemanticHighlightingCalculator.java]: Another thing: checkCanceled will throw an exception eventually that needs to be handled somewhere.

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/highlighting/SemanticHighlightVisitor.xtend [hunk 2]
```diff
@@ -68,7 +69,7 @@ class SemanticHighlightVisitor implements HighlightVisitor {
 	
 	override visit(PsiElement element) {
 		if (element instanceof BaseXtextFile) 
-			highlightCalculator.provideHighlightingFor(element.resource, acceptor)
+			highlightCalculator.provideHighlightingFor(element.resource, acceptor, CancelIndicator.NullImpl)
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/highlighting/SemanticHighlightVisitor.xtend]: We don't have a CancelIndicator yet that is based on IntelliJs abstractions?

File: plugins/org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/syntaxcoloring/SemanticHighlightingCalculator.java [hunk 3]
```diff
@@ -42,12 +45,17 @@
...
 	@Override
-	protected void doProvideHighlightingFor(XtextResource resource, IHighlightedPositionAcceptor acceptor) {
-		super.doProvideHighlightingFor(resource, acceptor);
+	protected void doProvideHighlightingFor(XtextResource resource, IHighlightedPositionAcceptor acceptor,
+			CancelIndicator cancelIndicator) {
+		super.doProvideHighlightingFor(resource, acceptor, cancelIndicator);
 		Iterator<EObject> iter = EcoreUtil.getAllContents(resource, true);
 		Set<AbstractRule> calledRules = Sets.newHashSet();
 		while(iter.hasNext()) {
+			operationCanceledManager.checkCanceled(cancelIndicator);
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/syntaxcoloring/SemanticHighlightingCalculator.java]: Maybe do this only after each rule?

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/highlighting/SemanticHighlightVisitor.xtend (hunk 2 above)
Comment (spoenemann) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/highlighting/SemanticHighlightVisitor.xtend]: No idea _(haha)_

File: plugins/org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/syntaxcoloring/SemanticHighlightingCalculator.java (hunk 3 above)
Comment (spoenemann) [inline plugins/org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/syntaxcoloring/SemanticHighlightingCalculator.java]: Yes, I'll change that later

**Pull request merged**